workflow/controller/operator.go

```diff
@@ -2365,6 +2365,12 @@ func (woc *wfOperationCtx) markWorkflowPhase(ctx context.Context, phase wfv1.Wor
 				}
 			}
 			woc.updated = true
+		} else {
+			woc.log.Info("Marking workflow completing")
+			woc.log.WithFields(log.Fields{"fromPhase": woc.wf.Status.Phase, "toPhase": wfv1.WorkflowCompleting}).
+				Debug("taskresults of workflow are incomplete or still have daemon nodes, so can't mark workflow completed")
+			woc.wf.Status.Phase = wfv1.WorkflowCompleting
+			woc.updated = true
```

Comment:
just to note that there is a very tiny possibility that we could set the `woc.wf.Status.Phase` to Success/Failed/Error up on line 2299, and then the Controller could go down before it gets overwritten here but that may be a remote enough possibility that we not worry about it

Reply:
Well, I am kind of wondering if there's a bit of restructuring that can be done. The other thing I see is that we're setting these Events even though we haven't made it the official new Phase yet:
```
switch phase {
		case wfv1.WorkflowRunning:
			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeNormal, "WorkflowRunning", "Workflow Running")
		case wfv1.WorkflowSucceeded:
			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeNormal, "WorkflowSucceeded", "Workflow completed")
		case wfv1.WorkflowFailed, wfv1.WorkflowError:
			woc.eventRecorder.Event(woc.wf, apiv1.EventTypeWarning, "WorkflowFailed", message)
		}
```

Reply:
Would you mind making a change so that we only set `woc.wf.Status.Phase = phase` and the Events when we should be?

Reply:
In the event of controller death none of the changes to wf.Status.Phase will make it out of the controller until Status is persisted to kubernetes at the end of reconciliation, in [persistUpdates](https://github.com/argoproj/argo-workflows/blob/d805b7fa6c36eac63d3dc24a84f2348a48f66be7/workflow/controller/operator.go#L718).

Events are queued, so when they are exposed to the outside world is also uncertain in a controller death scenario.

It does feel like this function is quite cognitively heavy, so attempting to make it simpler to understand is useful though.